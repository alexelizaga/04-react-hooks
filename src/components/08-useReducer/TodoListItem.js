import React from 'react';
import PropTypes from 'prop-types';

export const TodoListItem = ({ index, todo, handleToggle, handleDelete }) => {
  return (
    <li key={todo.id} className="list-group-item">
      <p
        className={`${todo.done && "complete"}`}
        onClick={() => handleToggle(todo.id)}
      >
        {index + 1}. {todo.desc}
      </p>
      <button
        className="btn btn-danger"
        onClick={() => handleDelete(todo.id)}
      >
        Borrar
      </button>
    </li>
  )
};

TodoListItem.propTypes = {
  index: PropTypes.number.isRequired,
  todo: PropTypes.object.isRequired,
  handleToggle: PropTypes.func.isRequired,
  handleDelete: PropTypes.func.isRequired,
}
