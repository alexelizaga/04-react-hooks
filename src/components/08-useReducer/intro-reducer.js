
const initialState = [{
  id: 1,
  todo: 'Comprar pan',
  done: false,
}];

const todoReducer = ( state = initialState, action ) => {
  switch (action?.type) {
    case 'add':
      return [ ...state, action.payload ]
  
    default:
      return state;
  }
}
let todos = todoReducer();

const newTodo = {
  id: 2,
  todo: 'Comprar leche',
  done: false,
}

const action = { type: 'add', payload: newTodo }

todos = todoReducer( todos, action );






console.log(todos);