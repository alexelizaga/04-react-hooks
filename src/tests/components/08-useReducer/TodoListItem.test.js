import '@testing-library/jest-dom';
import renderer from 'react-test-renderer';
import { fireEvent, render, screen } from '@testing-library/react';

import { TodoListItem } from '../../../components/08-useReducer/TodoListItem';
import { demoTodos } from '../../fixtures/demoTodos';

describe('Test <TodoListItem />', () => {

  const handleToggle = jest.fn();
  const handleDelete = jest.fn();

  const props = {
    index: 0,
    todo: demoTodos[0],
    handleToggle,
    handleDelete,
  }
  

  test('should render correctly', () => {
    const tree = renderer.create(<TodoListItem {...props} />).toJSON();
    expect(tree).toMatchSnapshot();
  });

  test('should call handleDelete', () => {
    render(<TodoListItem {...props} />);

    const btn = screen.getByTestId('TodoListItemBtn');
    fireEvent.click(btn);

    expect(handleDelete).toHaveBeenCalledWith(props.todo.id);
  });

  test('should show ToDo desc', () => {
    render(<TodoListItem {...props} />);

    const p = screen.getByTestId('TodoListItemDesc');

    expect(p.textContent.trim()).toBe(`${props.index + 1}. ${props.todo.desc}`);
  });

  test('should have complete className', () => {
    props.todo.done = true;
    render(<TodoListItem {...props} />);

    const p = screen.getByTestId('TodoListItemDesc');

    expect(p.className.includes('complete')).toBe(true);
  });

});