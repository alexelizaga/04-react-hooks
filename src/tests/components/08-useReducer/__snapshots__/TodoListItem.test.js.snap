// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Test <TodoListItem /> should render correctly 1`] = `
<li
  className="list-group-item"
  data-testid="TodoListItem"
>
  <p
    className="false"
    data-testid="TodoListItemDesc"
    onClick={[Function]}
  >
    1
    . 
    Aprender React
  </p>
  <button
    className="btn btn-danger"
    data-testid="TodoListItemBtn"
    onClick={[Function]}
  >
    Borrar
  </button>
</li>
`;
